kvbc/include/v4blockchain/detail/keys_history.h

```diff
@@ -0,0 +1,85 @@
+#include "categorization/updates.h"
+#include "v4blockchain/detail/categories.h"
+#include <rocksdb/compaction_filter.h>
+
+/*
+ * TODO Questions (efrat):
+ * 1. Filter condition: bigger than genesis? last_block - X blocks? if second option, last_block should be static,
+ *    atomic? and MAX_BLOCKS_NUM_SAVE_KEYS is TBD
+ * 2. revert: always of only 1 block?
+ * 3. a key can appear once in a single block?
+ * 4. test revertLastBlockKeys: would be better to add more blocks & assert after the revert?
+ * 5. all values in CF are empty string
+ * 6. rename getSnapshotValue : get key and snapshot (version), return the block_id where the key got it's value at
+ *    given snapshot. so, maybe getUpdateVersionAtSnapshot ?
+ * 7. should rename MAX_BLOCKS_NUM_SAVE_KEYS?
+ *
+ * */
+
+namespace concord::kvbc::v4blockchain::detail {
+
+class KeysHistory {
+ public:
+  KeysHistory(const std::shared_ptr<concord::storage::rocksdb::NativeClient>&,
+              const std::optional<std::map<std::string, concord::kvbc::categorization::CATEGORY_TYPE>>&);
+  void addBlockKeys(const concord::kvbc::categorization::Updates&, const BlockId, storage::rocksdb::NativeWriteBatch&);
+
+  // get the value and return deserialized value if needed.
+  std::optional<BlockId> getSnapshotValue(const std::string& category_id,
+                                          const std::string& key,
+                                          const BlockId version) const;
+
+  // Delete the last added block keys
+  void revertLastBlockKeys(const concord::kvbc::categorization::Updates&,
+                           const BlockId,
+                           storage::rocksdb::NativeWriteBatch&);
+
+ private:
+  void handleUpdatesImp(const BlockId block_id,
+                        const std::string& category_id,
+                        const std::vector<std::string>& updates_kv,
+                        concord::storage::rocksdb::NativeWriteBatch& write_batch);
+
+  void revertCategoryKeysImp(const std::string& category_id,
+                             const std::vector<std::string>& updates,
+                             const BlockId,
+                             concord::storage::rocksdb::NativeWriteBatch& write_batch);
+
+  struct KHCompactionFilter : ::rocksdb::CompactionFilter {
+    static ::rocksdb::CompactionFilter* getFilter() {
+      static KHCompactionFilter instance;
+      return &instance;
+    }
+    KHCompactionFilter() {}
+    const char* Name() const override { return "KeysHistoryCompactionFilter"; }
+    bool Filter(int /*level*/,
+                const ::rocksdb::Slice& key,
+                const ::rocksdb::Slice& val,
+                std::string* /*new_value*/,
+                bool* /*value_changed*/) const override;
+  };
+
+ private:
+  std::shared_ptr<concord::storage::rocksdb::NativeClient> native_client_;
+  v4blockchain::detail::Categories category_mapping_;
+  static BlockId last_block_;                            // TODO (efrat): should be atomic?
```

Comment:
It should be atomic since the compaction filter operates on a different thread.
Do you initialize it on start?
I recall having issues with data members and the compaction filter registration (you need to register it on RocksDB init); I suggest for symmetry defining  std::atomic<BlockId> Blockchain::global_latest_block_id = INVALID_BLOCK_ID;
and handling it as the global_genesis_block_id.

Reply:
Done